pkg/controller/utils/nutainx.go

```diff
@@ -0,0 +1,373 @@
+//
+// Parameters:
+//   - hiveFailureDomains: A slice of Hive failure domains (nutanix.FailureDomain) to convert.
+//
+// Returns:
+//   - []nutanixinstaller.FailureDomain: A slice of converted Installer failure domains (nutanixinstaller.FailureDomain).
+//   - []nutanixinstaller.PrismElement:  A slice of unique converted Installer PrismElements (nutanixinstaller.PrismElement).
+//   - []string:                       A slice of unique subnet UUIDs gathered from all Hive failure domains.
+func ConvertHiveFailureDomains(hiveFailureDomains []nutanix.FailureDomain) ([]nutanixinstaller.FailureDomain, []nutanixinstaller.PrismElement, []string) {
+	return convertFailureDomains[nutanix.FailureDomain, nutanixinstaller.FailureDomain, nutanixinstaller.PrismElement, nutanixinstaller.StorageResourceReference](
+		hiveFailureDomains,
+		convertHiveToInstallerPrismElement,
+		convertHiveToInstallerStorageResource,
+		getHiveSubnetUUIDs,
+		convertHiveToInstallerFailureDomain,
+	)
+}
+
+// ConvertInstallerFailureDomains converts Installer failure domains to Hive failure domains and returns unique PrismElements and SubnetUUIDs.
+//
+// This function specializes the generic convertFailureDomains function to convert
+// Installer's nutanixinstaller.FailureDomain type to Hive's nutanix.FailureDomain type.
+// It uses specific conversion functions for PrismElement, StorageResource, and FailureDomain
+// tailored for the Installer to Hive conversion. It also returns lists of unique PrismElements and SubnetUUIDs.
+//
+// Parameters:
+//   - installerFailureDomains: A slice of Installer failure domains (nutanixinstaller.FailureDomain) to convert.
+//
+// Returns:
+//   - []nutanix.FailureDomain:    A slice of converted Hive failure domains (nutanix.FailureDomain).
+//   - []nutanix.PrismElement:     A slice of unique converted Hive PrismElements (nutanix.PrismElement).
+//   - []string:                  A slice of unique subnet UUIDs gathered from all Installer failure domains.
+func ConvertInstallerFailureDomains(installerFailureDomains []nutanixinstaller.FailureDomain) ([]nutanix.FailureDomain, []nutanix.PrismElement, []string) {
+	return convertFailureDomains[nutanixinstaller.FailureDomain, nutanix.FailureDomain, nutanix.PrismElement, nutanix.StorageResourceReference](
+		installerFailureDomains,
+		convertInstallerToHivePrismElement,
+		convertInstallerToHiveStorageResource,
+		getInstallerSubnetUUIDs,
+		convertInstallerToHiveFailureDomain,
+	)
+}
+
+// convertHiveToInstallerPrismElement converts a Hive PrismElement to an Installer PrismElement.
+// Returns:
+//   - nutanixinstaller.PrismElement: The converted Installer PrismElement.
+func convertHiveToInstallerPrismElement(failureDomain nutanix.FailureDomain) (nutanixinstaller.PrismElement, error) {
+	if failureDomain.PrismElement.UUID == "" {
+		return nutanixinstaller.PrismElement{}, errors.New("no prism element found for failure domain " + failureDomain.Name)
```

Comment:
For debuggability, suggest making the three copies of this error message slightly different. E.g. this one could be, "no prism element found for hive failure domain".